File: untitled/src/StableMarriage.py


# Initialization
Men = ['Victor', 'Wyatt', 'Xavier', 'Yancey','Zeus']
Women = ['Amy', 'Bertha', 'Clare', 'Diane','Erika']

# Preferences
Men_Pref = {  # indicates the preferences of the men
    'Victor':   ['Amy', 'Bertha', 'Clare', 'Diane','Erika'],
    'Wyatt': ['Bertha', 'Clare', 'Diane', 'Amy','Erika'],
    'Xavier':  ['Clare', 'Diane', 'Amy', 'Bertha','Erika'],
    'Yancey':  ['Diane', 'Amy', 'Bertha', 'Clare','Erika'],
    'Zeus':  ['Amy', 'Bertha', 'Clare', 'Diane','Erika']
}

Women_Pref = {  # indicates the preferences of the women
    'Amy':  ['Wyatt', 'Xavier', 'Yancey', 'Zeus','Victor'],
    'Bertha':   ['Xavier', 'Yancey', 'Zeus', 'Victor','Wyatt'],
    'Clare': ['Yancey', 'Zeus', 'Victor', 'Wyatt','Xavier'],
    'Dianne':  ['Zeus', 'Victor', 'Wyatt', 'Xavier','Yancey'],
    'Erika': ['Victor', 'Wyatt', 'Xavier', 'Yancey','Zeus']
}
def main():
    Men_Free = list(Men)
    Women_Free = list(Women)

    # Part 3: Proposal
    Matches = {
        'Victor':  '',
        'Wyatt': '',
        'Xavier': '',
        'Yancey': '',
        'Zeus':''
        }
    key_list = list(Matches.keys())

    # the algorithm
    iteration = 0;
    while len(Men_Free) > 0:
        for man in key_list:
            print("-----Iteration ---- " ,iteration)
            iteration = iteration +1
            for woman in Men_Pref[man]:
                if woman not in list(Matches.values()):
                    Matches[man] = woman
                    Men_Free.remove(man)
                    print('{} is no longer a free man and is tentatively engaged to {} !'.format(man, woman))
                    break
                elif woman in list(Matches.values()):
                    current_suitor = list(Matches.keys())[list(Matches.values()).index(woman)]
                    w_list = Women_Pref.get(woman)
                    if w_list.index(man) < w_list.index(current_suitor):
                        Matches[man] = woman
                        print("men free->" , Men_Free)
                        print("man->>",man)
                        print("Current Sit -->",current_suitor)
                        Men_Free.remove(man)
                        Matches[current_suitor] = ''
                        Men_Free.append(current_suitor)
                        print('{} was earlier engaged to {} but now is engaged to {}! '.format(woman, current_suitor, man))
                        Men_Free.append(current_suitor)

    print('\n ')
    print('Stable Matching Finished ! Happy engagement !')
    for man in Matches.keys():
        print('{} is engaged to {} !'.format(man, Matches[man]))


if __name__ == "__main__":
    main()

File: untitled/src/StableMarriage.java
public class StableMarriage
{
    private int N, engagedCount;
    private String[][] menPref;
    private String[][] womenPref;
    private String[] men;
    private String[] women;
    private String[] womenPartner;
    private boolean[] menEngaged;
 
    /** Constructor **/
    public StableMarriage(String[] m, String[] w, String[][] mp, String[][] wp)
    {
        N = mp.length;
        engagedCount = 0;
        men = m;
        women = w;
        menPref = mp;
        womenPref = wp;
        menEngaged = new boolean[N];
        womenPartner = new String[N];
        calcMatches();
    }
    /** function to calculate all matches **/
    private void calcMatches()
    {
        int iteration = 1;
        while (engagedCount < N)
        {
            int free;
            for (free = 0; free < N; free++)
                if (!menEngaged[free])
                    break;
 
            for (int i = 0; i < N && !menEngaged[free]; i++)
            {
                int index = womenIndexOf(menPref[free][i]);
                if (womenPartner[index] == null)
                {
                    womenPartner[index] = men[free];
                    menEngaged[free] = true;
                    engagedCount++;
                    System.out.println(men[free] + " is temporarily matched with " + women[index]);
                }
                else
                {
                    String currentPartner = womenPartner[index];
                    if (morePreference(currentPartner, men[free], index))
                    {
                        System.out.println(men[free] + " proposes and " + women[index] + " who was earlier engaged to " + currentPartner + 
                                    " but now chooses " + men[free] + " as her new partner.");
                        womenPartner[index] = men[free];
                        menEngaged[free] = true;
                        menEngaged[menIndexOf(currentPartner)] = false;
                    }else {
                        System.out.println(men[free] + " proposes but " + women[index] + " prefers her current partner " + currentPartner + " over " + men[free]);
                    }
                }
            }            
        }
        printCouples();
    }
    /** function to check if women prefers new partner over old assigned partner **/
    private boolean morePreference(String curPartner, String newPartner, int index)
    {
        for (int i = 0; i < N; i++)
        {
            if (womenPref[index][i].equals(newPartner))
                return true;
            if (womenPref[index][i].equals(curPartner))
                return false;
        }
        return false;
    }
    /** get men index **/
    private int menIndexOf(String str)
    {
        for (int i = 0; i < N; i++)
            if (men[i].equals(str))
                return i;
        return -1;
    }
    /** get women index **/
    private int womenIndexOf(String str)
    {
        for (int i = 0; i < N; i++)
            if (women[i].equals(str))
                return i;
        return -1;
    }
    /** print couples **/
    public void printCouples()
    {
        System.out.println();
        System.out.println("Couples are : ");
        for (int i = 0; i < N; i++)
        {
            System.out.println(womenPartner[i] +" "+ women[i]);
        }
    }
    /** main function **/
    public static void main(String[] args) 
    {
        /** list of men **/
        String[] m ={"Victor", "Wyatt", "Xavier", "Yancey", "Zeus"} ;
        /** list of women **/
        String[] w ={"Amy", "Bertha", "Clare", "Diane", "Erika"};;
 
        /** men preference **/
        String[][] mp = {
            {"Amy", "Bertha", "Clare", "Diane", "Erika"},  // Victor's preferences
            {"Bertha", "Clare", "Diane", "Amy", "Erika"},  // Wyatt's preferences
            {"Clare", "Diane", "Amy", "Bertha", "Erika"},  // Xavier's preferences
            {"Diane", "Amy", "Bertha", "Clare", "Erika"},  // Yancey's preferences
            {"Amy", "Bertha", "Clare", "Diane", "Erika"}   // Zeus's preferences
        };
        /** women preference **/                      
        String[][] wp = {
            {"Wyatt", "Xavier", "Yancey", "Zeus", "Victor"},   // Amy's preferences
            {"Xavier", "Yancey", "Zeus", "Victor", "Wyatt"},   // Bertha's preferences
            {"Yancey", "Zeus", "Victor", "Wyatt", "Xavier"},   // Clare's preferences
            {"Zeus", "Victor", "Wyatt", "Xavier", "Yancey"},   // Dianne's preferences
            {"Victor", "Wyatt", "Xavier", "Yancey", "Zeus"}    // Erika's preferences
        };
 
        StableMarriage gs = new StableMarriage(m, w, mp, wp);                        
    }
}